edalize/nextpnr.py

```diff
@@ -0,0 +1,124 @@
+                    {
+                        "name": "output_format",
+                        "type": "String",
+                        "desc": "Output file format. Legal values are *fasm*",
+                    },
+                    {
+                        "name": "nextpnr_as_subtool",
+                        "type": "bool",
+                        "desc": "Determines if nextpnr is run as a part of bigger toolchain, or as a standalone tool",
+                    },
+                ],
+                "lists": [
+                    {
+                        "name": "nextpnr_impl_options",
+                        "type": "String",
+                        "desc": "Additional options for the implementation command",
+                    },
+                ],
+            }
+
+            combined_members = nextpnr_help["members"]
+            combined_lists = nextpnr_help.get("lists", [])
+            yosys_members = yosys_help["members"]
+            yosys_lists = yosys_help["lists"]
+
+            combined_members.extend(
+                m
+                for m in yosys_members
+                if m["name"] not in [i["name"] for i in combined_members]
+            )
+            combined_lists.extend(
+                l
+                for l in yosys_lists
+                if l["name"] not in [i["name"] for i in combined_lists]
+            )
+
+            return {
+                "description": "Open source Place and Route tool targeting many different FPGAs",
+                "members": combined_members,
+                "lists": combined_lists,
+            }
+
+    def configure_main(self):
+        (src_files, incdirs) = self._get_fileset_files()
+
+        yosys_synth_options = self.tool_options.get('yosys_synth_options', '')
+        yosys_edam = {
+                'files'         : self.files,
+                'name'          : self.name,
+                'toplevel'      : self.toplevel,
+                'parameters'    : self.parameters,
+                'tool_options'  : {'yosys' : {
+                                        'arch' : 'xilinx',
+                                        'yosys_synth_options' : yosys_synth_options,
+                                        'yosys_as_subtool' : True,
+                                        }
+                                }
+                }
+
+        yosys = getattr(import_module("edalize.yosys"), "Yosys")(
+            yosys_edam, self.work_root
+        )
+        yosys.configure()
+
+        part_of_toolchain = self.tool_options.get("nextpnr_as_subtool", False)
+
+        file_table = []
+
+        output_format = self.tool_options.get("output_format", "fasm")
+        arch = self.tool_options.get("arch", "xilinx")
+
+        synth_design = None
+        chipdb = None
+        xdc = None
+
+        for f in src_files:
+            if f.file_type in ["synthJson"]:
+                synth_design = f.name
+            elif f.file_type in ["bba"]:
```

Comment:
AFAIK `nextpnr` has embedded data base for e.g. ice40, which is much smaller than the xilinx chipdb. The chipdb can be either delivered by user or it can be used straight from e.g. conda package. In case of conda package, the `nextpnr-xilinx` from `symbiflow` channel has chipdb placed in `/path/to/miniconda/share/nextpnr-xilinx/xc7a35tcpg236-1.bin`.

Reply:
I see. Some follow-up questions then
1. Is there ever a reason for regular users to override the one that comes with symbiflow?
2. If we assume that users are fine with the one that symbiflow installs, can we automatically find out where it is installed?

I really prefer to hide these things from users. Especially when it comes down to paths. We don't want users to hard code a path in their FuseSoC core description files, and we don't want them to specify it on the command-line every time they run it either

Reply:
> 1. Is there ever a reason for regular users to override the one that comes with symbiflow?

I think regular users would use an official package rather than a custom one.

> 2. If we assume that users are fine with the one that symbiflow installs, can we automatically find out where it is installed?

We can check if chipdb file exists in `$CONDA_PREFIX/share/nextpnr-xilinx/` by default, if not then we can search for it in a custom path provided by an optional `chipDataBase` file. I think it makes sense to hide path related stuff from regular users, but on the other hand leaving an option for a custom chipdb path would be also great for developers.

Reply:
I added some related changes, please take a look: [link](https://github.com/olofk/edalize/pull/193/files#diff-f2231ab9d2b0eca263ed813cc4ba1ee5R96-R112)

Reply:
I like that idea! It hides it completely from regular users. One question about CONDA_PREFIX though. Am I supposed to define that myself, or does the toolchain set it somehow? I don't see it defined in the environment I'm using

Reply:
To be clear, I'm fine with using CONDA_PREFIX if that's the most sensible choice, but as I understand it, that's some general-purpose package manager. Do we know that nextpnr-xilinx is always installed with conda?

But if CONDA_PREFIX is the way to go, then I think we should also add a check to see if it's set before running anything (like the modelsim backend checks that $MODEL_TECH is set)

Reply:
`CONDA_PREFIX` env variable appears when you activate conda: `conda activate` (assuming you have `conda` in PATH)

Reply:
What happens if nextpnr-xilinx is not installed via conda?

Reply:
Hm, afaik there are two ways you can obtain `nextpnr-xilinx`:

1. Conda package (used by default in SymbiFlow)
2. Get it from sources (https://github.com/daveshah1/nextpnr-xilinx)

In both cases I think we should assume that `nextpnr-xilinx` binary is accessible from `PATH` (either it is installed with conda or by make install).

Chipdb files are build separately, take a look at [readme](https://github.com/daveshah1/nextpnr-xilinx#building-the-arty-example---xray-database). In case of conda, chipdb files are delivered along with a `nextpnr-xilinx` binary. If a user wants to use chipdbs built from sources, we'll need to handle it either with a separate env variable or we can assume that building chipdbs from sources is advanced enough approach that user should provide these files via `chipDataBase` entry in edam.

Reply:
As @kowalewskijan mentioned previously, the bba binary data files are quite large compared to the icestorm ones, making it hard to store all of them alongside with the nextpnr-xilinx binary.

IMO, for the time being we could leave the conda_prefix as the last resource to get the bba files, unless specified by the user.
Therefore, the user can have two options (for now):
- install nextpnr-xilinx through the SymbiFlow/anaconda packages
- locally generate the binary data files and provide them to edalize.